solidity/test/TokenTimeWeightedAverage.js

```diff
@@ -0,0 +1,502 @@
+const { accounts, defaultSender, contract, web3 } = require('@openzeppelin/test-environment');
+const { expectRevert, expectEvent, constants, BN, time } = require('@openzeppelin/test-helpers');
+const { expect } = require('../../chai-local');
+
+const Decimal = require('decimal.js');
+
+const { ZERO_ADDRESS } = constants;
+const { latest, duration } = time;
+
+const TokenTimeWeightedAverage = contract.fromArtifact('TestTokenTimeWeightedAverage');
+
+const ROLE_OWNER = web3.utils.keccak256('ROLE_OWNER');
+const ROLE_SEEDER = web3.utils.keccak256('ROLE_SEEDER');
+
+const initTWA = (start) => {
+    return {
+        firstSampleTime: Decimal(start.toString()),
+        lastSampleTime: Decimal(start.toString()),
+        accumulators: {
+            [start]: Decimal(0)
+        },
+        lastAccumulator: Decimal(0),
+        lastAccumulatorTime: Decimal(0)
+    };
+};
+
+const addTWASample = (acc, n, d, time) => {
+    const sampleTime = time.toNumber();
```

Comment:
Not sure what the type of `time` is in this function, and why you chose to convert it to `Number`.
But since you're converting it to `Decimal` below, it seems pointless to convert it to `Number` here.
For example, what if it represents an integer value larger than `Number.MAX_SAFE_INTEGER`?
If the issue is that `time` is a `BN` object, then the typical solution is:
```
const sampleTime = Decimal(time.toString());
```

Reply:
👍